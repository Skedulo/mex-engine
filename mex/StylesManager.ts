import ThemeManager from './colors/ThemeManager'
import {Platform} from "react-native";

type StyleConstant = {
    defaultHorizontalPadding: number
    defaultVerticalPadding: number,
    smallVerticalPadding: number,
    componentVerticalPadding: number,
    betweenTextSpacing: number,
    captionTextSize: number,
    betweenComponentVerticalSpacing: number
}

class StylesManager {
    styles: any;
    stylesConstant: StyleConstant = {
        defaultHorizontalPadding: 16,
        defaultVerticalPadding: 16,
        smallVerticalPadding: 8,
        componentVerticalPadding: 12,
        betweenTextSpacing: 6,
        captionTextSize: 12,
        betweenComponentVerticalSpacing: 32
    };

    initializeStyles() {
        let colors = ThemeManager.getColorSet()

       this.styles =  {
           text: {
               "color": colors.skeduloText,
               "fontSize": 20,
           },

           errorText: {
               "color": colors.red800,
               "fontSize": 14,
           },
           textRegular: {
               fontWeight: "normal",
               fontSize: 16,
               flexWrap: 'wrap',
               color: colors.skeduloText
           },
           textTitleListItem: {
               fontWeight: "600",
               fontSize: 16,
               color: colors.skeduloText
           },
           textCaptionListItem: {
               fontWeight: "normal",
               fontSize: 16,
               color: colors.navy600
           },
           textCaptionSelector: {
               fontWeight: "normal",
               fontSize: 14,
               color: colors.navy600
           },
           textCaption: {
               fontWeight: "normal",
               fontSize: 12,
               flexWrap: 'wrap',
               color: colors.skeduloText
           },
           textMedium: {
               fontWeight: "500",
               fontSize: 16,
               color: colors.skeduloText
           },

           textHeadingBold: {
               fontWeight: "bold",
               fontSize: 18,
               color: colors.skeduloText
           },

           navigationTitle: {
               fontSize: 16,
               fontWeight: "bold"
           },

           navigationSubtitle: {
               fontSize: 14
           },
           editText: {
               borderColor: colors.navy300,
               borderWidth: 1,
               borderRadius: 2,
               marginTop: 8,
               height: 48,
               fontSize: 16,
               paddingHorizontal: 16,
               justifyContent: 'center',
               color: colors.skeduloText,
           },
           selector: {
               borderWidth: 1,
               borderRadius: 2,
               marginTop: 8,
               height: 48,
               paddingHorizontal: 16,
           },
           searchBarTextContainer: {
               borderColor: colors.navy100,
               backgroundColor: colors.navy75,
               borderWidth: 1,
               borderRadius: 10,
               paddingHorizontal: 12,
           },
           searchBarText: {
               backgroundColor: "transparent",
               borderWidth: 0,
               borderRadius: 0,
               paddingHorizontal: 12,
               paddingTop: 12,
               paddingBottom: 12,
               color: colors.skeduloText
           },
           button: {
               backgroundColor: colors.skeduloBlue,
               height: 44,
               borderRadius: 8,
               paddingVertical: 10,
               paddingHorizontal: 12,
               justifyContent: "center"
           },
           buttonSmall: {
               backgroundColor: colors.skeduloBlue,
               borderRadius: 8,
               justifyContent: "center",
               height: 32,
               paddingVertical: 8,
               paddingHorizontal: 20
           },
           buttonDefault: {
               height: 44,
               borderRadius: 8,
               paddingVertical: 10,
               paddingHorizontal: 12,
               justifyContent: "center",
               borderColor: colors.skeduloBlue,
               borderWidth: 1
           },
           headerButtonText: {
               fontSize: 16,
               color: colors.white,
               fontWeight: 'normal',
               alignSelf: 'center',
           },
           buttonInsideText: {
               fontSize: 16,
               color: colors.white,
               fontWeight: "600",
               alignSelf: "center",
           },
           buttonSmallInsideText: {
               fontSize: 14,
               color: colors.white,
               fontWeight: "600",
               alignSelf: "center",
           },
           buttonDefaultInsideText: {
               fontSize: 16,
               color: colors.skeduloBlue,
               fontWeight: "500",
               alignSelf: "center",
           },

           buttonLinkText: {
               fontSize: 16,
               color: colors.skeduloText,
               fontWeight: "500",
               alignSelf: "center",
           },

           headerButton: {
               color: colors.white
           },
           headerButtonAlt: {
               color: colors.skedBlue800
           },
           checkBoxProps: {
               size: 25,
               useNativeDriver: true,
               unfillColor: colors.white,
               disableBuiltInState: true,
               iconStyle: {
                   borderRadius: Platform.OS === 'ios' ? 13 : 5
               }
           },
           headerBarSelectView: {
               borderRadius: 8,
               flexDirection: 'row',
               justifyContent:'space-between',
               paddingTop: 19,
               paddingBottom:15,
               paddingHorizontal: 16
           }
       }
    }

    getStyles() { return this.styles }
    getStyleConst() : StyleConstant { return this.stylesConstant }

}

export default new StylesManager() as StylesManager

