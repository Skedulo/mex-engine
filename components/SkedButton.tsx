import React, {useState} from "react";
import {Text, TouchableWithoutFeedback, View} from "react-native";
import ThemeManager from "../mex/colors/ThemeManager";
import StylesManager from "../mex/StylesManager";
import {PromiseFn} from "react-async";
import MexAsyncText from "./MexAsyncText";
import {SkedButtonTheme} from "@skedulo/mex-types";

export enum SkedButtonSize {
    SMALL = 'small',
    LARGE = 'large',
}

type Props = {
    textPromiseFn: PromiseFn<string>
    onPress?: (() => void) | undefined;
    theme?: SkedButtonTheme,
    disabled?: boolean | undefined
    size?: SkedButtonSize
}

const buttonStyleConfig = {
    primary: {
        backgroundColor: ThemeManager.getColorSet().skedBlue800,
        pressedBackgroundColor: ThemeManager.getColorSet().skedBlue900,
        disabledBackgroundColor: ThemeManager.getColorSet().skedBlue200,
        textColor: ThemeManager.getColorSet().white,
        disabledTextColor: ThemeManager.getColorSet().white,
    },
    success: {
        backgroundColor: ThemeManager.getColorSet().green800,
        pressedBackgroundColor: ThemeManager.getColorSet().green900,
        disabledBackgroundColor: ThemeManager.getColorSet().green200,
        textColor: ThemeManager.getColorSet().white,
        disabledTextColor: ThemeManager.getColorSet().white,
    },
    default: {
        backgroundColor: ThemeManager.getColorSet().navy50,
        pressedBackgroundColor: ThemeManager.getColorSet().navy100,
        disabledBackgroundColor: ThemeManager.getColorSet().navy50,
        textColor: ThemeManager.getColorSet().skeduloText,
        disabledTextColor: ThemeManager.getColorSet().navy300,
    }
}

const SkedButton: React.FC<Props> = ({textPromiseFn, onPress, disabled = false, theme, size }: Props) => {
    const buttonTheme = theme || 'success'
    const buttonSize = size || SkedButtonSize.LARGE
    const [isPressing, setIsPressing] = useState(false)

    const styles = StylesManager.getStyles()

    const onButtonPressed = () => {
        onButtonPressedOut()
        onPress?.()
    }
    const onButtonFocused = () => {
        setIsPressing(true)
    }

    const onButtonPressedOut = () => {
        setIsPressing(false)
    }

    const getBackgroundColor = () => {
        if (disabled) {
            return buttonStyleConfig[buttonTheme].disabledBackgroundColor
        }
        if (isPressing) {
            return buttonStyleConfig[buttonTheme].pressedBackgroundColor
        }
        return buttonStyleConfig[buttonTheme].backgroundColor
    }

    return (
        <TouchableWithoutFeedback
            testID="sked-button"
            disabled={disabled}
            onPressOut={onButtonPressedOut}
            onPressIn={onButtonFocused}
            onPress={onButtonPressed}>
            <View
                style={[buttonSize === SkedButtonSize.SMALL ? styles.buttonSmall : styles.button, {backgroundColor: getBackgroundColor()}]}>
                <MexAsyncText promiseFn={textPromiseFn}>
                    {(text) =>
                        <Text style={[buttonSize === SkedButtonSize.SMALL ? styles.buttonSmallInsideText : styles.buttonInsideText,
                            {color:  disabled ? buttonStyleConfig[buttonTheme].disabledTextColor : buttonStyleConfig[buttonTheme].textColor}]}>{text}</Text>
                    }
                </MexAsyncText>
            </View>
        </TouchableWithoutFeedback>)
}

export default SkedButton
