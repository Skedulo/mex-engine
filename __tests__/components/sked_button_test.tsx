import React from 'react';
import SkedButton from "../../components/SkedButton";
import {act, fireEvent, render} from "@testing-library/react-native";
import StylesManager from "../../mex/StylesManager";

describe('SkedButton Component Tests', function () {

    it('SkedButton basic renderer', async function () {
        StylesManager.initializeStyles()

        const getText = async (): Promise<string> => {
            return Promise.resolve("Test button")
        }

        const tree = render(<SkedButton textPromiseFn={getText} />)

        await act(async () => {
            await new Promise((r) => setTimeout(r, 10));
        });

        expect(tree.toJSON()).toMatchSnapshot()

        expect(tree.getByText("Test button")).toBeTruthy()

        fireEvent(tree.getByTestId("sked-button"), "onPressIn")

        expect(tree.toJSON()).toMatchSnapshot()

        fireEvent(tree.getByTestId("sked-button"), "onPressOut")

        expect(tree.toJSON()).toMatchSnapshot()
    });
});
