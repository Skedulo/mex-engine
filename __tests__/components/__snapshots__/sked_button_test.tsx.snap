// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SkedButton Component Tests SkedButton basic renderer 1`] = `
<View
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#0461cb",
        "borderRadius": 8,
        "height": 44,
        "justifyContent": "center",
        "paddingHorizontal": 12,
        "paddingVertical": 10,
      },
      Object {
        "backgroundColor": "#03AA55",
      },
    ]
  }
  testID="sked-button"
>
  <Text
    style={
      Array [
        Object {
          "alignSelf": "center",
          "color": "#ffffff",
          "fontSize": 16,
          "fontWeight": "600",
        },
        Object {
          "color": "#ffffff",
        },
      ]
    }
  >
    Test button
  </Text>
</View>
`;

exports[`SkedButton Component Tests SkedButton basic renderer 2`] = `
<View
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#0461cb",
        "borderRadius": 8,
        "height": 44,
        "justifyContent": "center",
        "paddingHorizontal": 12,
        "paddingVertical": 10,
      },
      Object {
        "backgroundColor": "#028844",
      },
    ]
  }
  testID="sked-button"
>
  <Text
    style={
      Array [
        Object {
          "alignSelf": "center",
          "color": "#ffffff",
          "fontSize": 16,
          "fontWeight": "600",
        },
        Object {
          "color": "#ffffff",
        },
      ]
    }
  >
    Test button
  </Text>
</View>
`;

exports[`SkedButton Component Tests SkedButton basic renderer 3`] = `
<View
  accessibilityState={
    Object {
      "disabled": false,
    }
  }
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "backgroundColor": "#0461cb",
        "borderRadius": 8,
        "height": 44,
        "justifyContent": "center",
        "paddingHorizontal": 12,
        "paddingVertical": 10,
      },
      Object {
        "backgroundColor": "#03AA55",
      },
    ]
  }
  testID="sked-button"
>
  <Text
    style={
      Array [
        Object {
          "alignSelf": "center",
          "color": "#ffffff",
          "fontSize": 16,
          "fontWeight": "600",
        },
        Object {
          "color": "#ffffff",
        },
      ]
    }
  >
    Test button
  </Text>
</View>
`;
